File: app/src/main/java/com/aar/app/wsp/di/modules/DataSourceModule.java
package com.aar.app.wsp.di.modules;

import android.content.Context;

import com.aar.app.wsp.data.room.GameDatabase;
import com.aar.app.wsp.data.room.GameThemeDataSource;
import com.aar.app.wsp.data.room.UsedWordDataSource;
import com.aar.app.wsp.data.room.WordDataSource;
import com.aar.app.wsp.data.sqlite.DbHelper;
import com.aar.app.wsp.data.sqlite.GameDataSource;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;

/**
 * Created by abdularis on 18/07/17.
 */

@Module
public class DataSourceModule {

    @Provides
    @Singleton
    GameDatabase provideGameDatabase(Context context) {
        return GameDatabase.getInstance(context);
    }

    @Provides
    @Singleton
    DbHelper provideDbHelper(Context context) {
        return new DbHelper(context);
    }

    @Provides
    @Singleton
    GameDataSource provideGameRoundDataSource(DbHelper dbHelper, UsedWordDataSource usedWordDataSource) {
        return new GameDataSource(dbHelper, usedWordDataSource);
    }

    @Provides
    @Singleton
    GameThemeDataSource provideGameThemeDataSource(GameDatabase gameDatabase) {
        return gameDatabase.getGameThemeDataSource();
    }

    @Provides
    @Singleton
    WordDataSource provideWordDataSource(GameDatabase gameDatabase) {
        return gameDatabase.getWordDataSource();
    }

    @Provides
    @Singleton
    UsedWordDataSource provideUsedWordDataSource(GameDatabase gameDatabase) {
        return gameDatabase.getUsedWordDataSource();
    }
}

File: app/src/main/java/com/aar/app/wsp/di/modules/DataSourceModule.kt
package com.aar.app.wsp.di.modules

import android.content.Context
import com.aar.app.wsp.data.room.GameDatabase
import com.aar.app.wsp.data.room.GameThemeDataSource
import com.aar.app.wsp.data.room.UsedWordDataSource
import com.aar.app.wsp.data.room.WordDataSource
import com.aar.app.wsp.data.sqlite.DbHelper
import com.aar.app.wsp.data.sqlite.GameDataSource
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

/**
 * Created by abdularis on 18/07/17.
 */
@Module
class DataSourceModule {
    @Provides
    @Singleton
    fun provideGameDatabase(context: Context): GameDatabase {
        return GameDatabase.getInstance(context)
    }

    @Provides
    @Singleton
    fun provideDbHelper(context: Context): DbHelper {
        return DbHelper(context)
    }

    @Provides
    @Singleton
    fun provideGameRoundDataSource(
        dbHelper: DbHelper,
        usedWordDataSource: UsedWordDataSource
    ): GameDataSource {
        return GameDataSource(dbHelper, usedWordDataSource)
    }

    @Provides
    @Singleton
    fun provideGameThemeDataSource(gameDatabase: GameDatabase): GameThemeDataSource {
        return gameDatabase.gameThemeDataSource
    }

    @Provides
    @Singleton
    fun provideWordDataSource(gameDatabase: GameDatabase): WordDataSource {
        return gameDatabase.wordDataSource
    }

    @Provides
    @Singleton
    fun provideUsedWordDataSource(gameDatabase: GameDatabase): UsedWordDataSource {
        return gameDatabase.usedWordDataSource
    }
}